### ISSUE TITLE
`ignore_label` argument in Caffe2 `SoftmaxWithLoss`

### ISSUE THREAD

**Issue by Feynman27** [pytorch/pytorch#12576]
## 🚀 Feature

Similar to the Pytorch implementation of [crossentropyloss](https://pytorch.org/docs/stable/nn.html#crossentropyloss), it'd be nice to have an `ignore_index` or `ignore_label` argument in the Caffe2 [SoftmaxWithLoss](https://github.com/pytorch/pytorch/blob/a4120fa132849028d2c84dee684b3dda1ef4e8b2/caffe2/operators/softmax_with_loss_op.cc#L134). 

## Motivation
In practice, I often experience the need to include additional loss terms (e.g. additional classification losses) that are applicable to some classes but not others. For example, say that I would like to predict gender for only the `person` class, but would like to ignore the `gender` classification for other classes. I could simply set the `ignore_label = -1` for classes that don't apply. I'm currently finding it difficult to do this in Caffe2.

**Comment by Feynman27** [comment]
I was able to train this model by modifying `softmax_ops.cu` to the following:

```
#define DONTCARE (-1)
__global__ void LabelCrossEntropyKernel(
    const int N,
    const int D,
    const float* logPdata,
    const int* labeldata,
    const float* weights,
    float* Ydata, 
    float* weight_data) {
  CUDA_1D_KERNEL_LOOP(i, N) {
    const int label = static_cast<int>(labeldata[i]);
    if (label != DONTCARE) {
      CUDA_KERNEL_ASSERT(label >= 0 && label < D);
      float weight = weights ? weights[i] : 1.0;
      Ydata[i] = -logPdata[i * D + label] * weight;
      weight_data[i] = weight;
    } else {
      Ydata[i] = 0;
      weight_data[i] = 0;
    }
  }
}

__global__ void LabelCrossEntropyGradientKernel(
    const int N,
    const int D,
    const float* Pdata,
    const int* labeldata,
    float* dXdata, 
    float* weights_) {
  CUDA_1D_KERNEL_LOOP(i, N) {
    const int label = static_cast<int>(labeldata[i]);
    if (label != DONTCARE) {
      int idx = i * D + label;
      dXdata[idx] = Pdata[idx] - 1.;
      weights_[i] = 1.0;
    } else {
      // Ignore-label, so set all gradients to zero
      for (int c = 0; c < D; ++c) {
        int idx = i * D + c;
        dXdata[idx] = 0.0;
      }
      weights_[i] = 0.0;
    }
  }
}
...
template<>
bool SoftmaxWithLossOp<float, CUDAContext>::RunOnDevice() {
  auto& X = Input(0);  // Logits
  auto& T = Input(1);  // Labels / targets
  auto* P = Output(0); // Probabilities from softmax
  auto* avg_loss = Output(1); // Average loss
  const float* weights = (InputSize() > 2 ? Input(2).data<float>() : NULL);

  const auto canonical_axis = X.canonical_axis_index(axis_);
  int N, D;
  N = X.size_to_dim(canonical_axis); // batch size
  D = X.size_from_dim(canonical_axis);
  P->ResizeLike(X);
  total_weight_ptr_.Resize(1);

  if (label_prob_mode_) {
    CAFFE_ENFORCE_GE(T.ndim(), 2);
    CAFFE_ENFORCE_EQ(T.size_to_dim(canonical_axis), N);
    CAFFE_ENFORCE_EQ(T.size_from_dim(canonical_axis), D);
  } else {
    if (T.ndim() == canonical_axis) {
      CAFFE_ENFORCE_EQ(T.size(), N);
    } else {
      CAFFE_ENFORCE_EQ(T.size_to_dim(canonical_axis), N);
      CAFFE_ENFORCE_EQ(T.size_from_dim(canonical_axis), 1);
    }
  }

  avg_loss->Resize(vector<TIndex>());
  if (losses_.size() != N) {
    losses_.Resize(N);
  }
  if (weights_.size() != N) {
    weights_.Resize(N);
  }
  if (rowmax_.size() != N) {
    rowmax_.Resize(N);
  }
  if (sum_multiplier_.size() != D) {
    sum_multiplier_.Resize(D);
    math::Set<float, CUDAContext>(
        D, 1.f, sum_multiplier_.mutable_data<float>(), &context_);
  }
  Softmax(
      N,
      D,
      X.data<float>(),
      sum_multiplier_.data<float>(),
      losses_.mutable_data<float>(),
      rowmax_.mutable_data<float>(),
      P->mutable_data<float>(),
      !label_prob_mode_, // logarithmic output
      &context_);
  // Compute label xent loss per example
  if (!label_prob_mode_) {
    LabelCrossEntropyKernel<<<
        CAFFE_GET_BLOCKS(N),
        CAFFE_CUDA_NUM_THREADS,
        0,
        context_.cuda_stream()>>>(
        N,
        D,
        P->data<float>(),
        T.data<int>(),
        weights,
        losses_.mutable_data<float>(),
        weights_.mutable_data<float>());
    // Since we had logarithmic output, we need to exponentiate
    // them again.
    math::Exp<float, CUDAContext>(
        N * D, P->data<float>(), P->mutable_data<float>(), &context_);
  } else {
    ProbCrossEntropyKernel<<<
        std::min(N, CAFFE_MAXIMUM_NUM_BLOCKS),
        CAFFE_CUDA_NUM_THREADS,
        0,
        context_.cuda_stream()>>>(
        N,
        D,
        P->data<float>(),
        T.data<float>(),
        weights,
        losses_.mutable_data<float>());
  }

  float total_weight = N;
  if (weights) {
    // Sum weights
    math::Sum<float, CUDAContext>(
        N, weights, total_weight_ptr_.mutable_data<float>(), &context_, &scratch_);
    CUDA_CHECK(cudaMemcpyAsync(
        &total_weight,
        total_weight_ptr_.data<float>(),
        sizeof(float),
        cudaMemcpyDeviceToHost,
        context_.cuda_stream()));
  } else {
    math::Sum<float, CUDAContext>(
        weights_.size(),
        weights_.data<float>(),
        total_weight_ptr_.mutable_data<float>(),
        &context_,
        &scratch_);
    CUDA_CHECK(cudaMemcpyAsync(
        &total_weight,
        total_weight_ptr_.data<float>(),
        sizeof(float),
        cudaMemcpyDeviceToHost,
        context_.cuda_stream()));
  } 
  // Sum of all losses
  float* avg_loss_data = avg_loss->mutable_data<float>();
  math::Sum<float, CUDAContext>(
      losses_.size(), losses_.data<float>(), avg_loss_data, &context_, &scratch_);
  // Average of input batch size
  if (total_weight > 0) {
    math::Scale<float, CUDAContext>(
        1, scale_ / total_weight, avg_loss_data, avg_loss_data, &context_);
  }

  return true;
}

```

I also added the unit test to `softmax_ops_test.py`:

```
    @given(n=st.integers(1,2), D=st.integers(2,3), **hu.gcs_gpu_only)
    def test_softmax_with_ignore_loss(self, n, D, gc, dc):
        # n = number of examples, D = |labels|
        # Initialize X and add 1e-2 for numerical stability
        np.random.seed(2603)
        X = np.random.rand(n, D).astype(np.float32)
        X = X + 1e-2

        # Initialize label. Some labels are (-1) i.e. "DONT CARE"
        label = (np.random.rand(n) * (D + 1)).astype(np.int32) - 1
        print(label)

        # Reference implementation of cross entropy with soft labels
        def label_softmax_crossent(X, label):
            probs = np.zeros((n, D))
            rowmax = np.zeros(n)
            for i in range(n):
                rowmax[i] = max(X[i, ])
                # We need to subtract the max to avoid numerical issues
                probs[i] = X[i] - rowmax[i]
                exps = np.exp(probs[i, ])
                norm = sum(exps)
                probs[i, ] = exps / norm

            label_xent = [-np.log(max(probs[i][label[i]], 1e-20))
                          for i in range(n)]
            #label_xent = [-np.log(max(probs[i][label[i]], 1e-20))
            #              if label[i]!=-1 else 1e-20 for i in range(n)]
            total_xent = 0.0
            total_weight = 0.0
            for i in range(n):
                l = label[i]
                if (l!=(-1)):
                    total_xent += -np.log(max(probs[i,l], 1e-20))
                    total_weight += 1.0
            #avgloss = np.sum(label_xent) / float(n)
            avgloss = total_xent / total_weight
            print(avgloss)
            return (probs, avgloss)

        op = core.CreateOperator(
            "SoftmaxWithLoss",
            ["X", "label"],
            ["probs", "avgloss"],
        )

        self.assertReferenceChecks(
            device_option=gc,
            op=op,
            inputs=[X, label],
            reference=label_softmax_crossent,
        )

        self.assertGradientChecks(
            gc, op, [X, label], 0, [1], stepsize=1e-4, threshold=1e-2)

    @given(n=st.sampled_from([0, 2, 4, 71, 103, 555, 751, 1201]),
           D=st.sampled_from([4, 8, 64, 79, 256, 333, 1000]),
           engine=st.sampled_from([None, 'CUDNN']),
           **hu.gcs)
```

However, at the moment, this only works on the GPU and breaks any tests with `label_prob` ( `def test_softmax_with_loss_axis_2` and `test_softmax_with_loss_label_prob`). All other tests pass.
